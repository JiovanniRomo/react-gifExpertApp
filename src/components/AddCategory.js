import React, { useState } from 'react'
import PropTypes from 'prop-types';

//Este componente solo nos ayuda a crear nuevas categorias

export const AddCategory = ({setCategory}) => {


    const [inputValue, setinputValue] = useState('');

    const handleInputCange = (e) => {
        setinputValue(e.target.value);
    }


    //Prevenimos el refresh del navegador
    const handleSubmit = (e) => {
        e.preventDefault();

        if(inputValue.trim().length > 2){
            setCategory(cats => [inputValue, ...cats]);
            setinputValue(''); //esto es para no hacer una doble inserci√≥n, borra lo que el usuario ya escribio
        }


    }

    return (
        <form onSubmit={handleSubmit}>
            <p>{inputValue}</p>
            <input
                type='text'
                value={inputValue}
                onChange={ handleInputCange }
            />
        </form>
    )
}

AddCategory.propTypes = {
    setCategory: PropTypes.func.isRequired
}
