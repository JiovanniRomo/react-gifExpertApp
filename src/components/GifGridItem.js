import React from 'react';
import PropTypes from 'prop-types';

//Este componente solo crea las cracteristicas de cada gif

export const GifGridItem = ({title, url}) => {

    return (
        <div className='card animate__animated animate__fadeInRight'>
            <img 
                src={url}
                alt={title}
            />
            <p>{title}</p>
        </div>
    )
}

GifGridItem.propType = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}
