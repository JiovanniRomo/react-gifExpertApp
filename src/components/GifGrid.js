import React from 'react';
import PropTypes from 'prop-types';
import { GifGridItem } from './GifGridItem';
import { useFetchGifs } from '../hooks/useFetchGifs';

//Este componente recopila todos los items para mostrarlos

export const GifGrid = ({category}) => {

    const {data:images, loading} = useFetchGifs(category);

    
    return (
        //Como ya estamos retornando 2 objetos(h3, div) , usamos fragment
        //! ES IMPORTANTE USAR CLASSNAME PARA DEFINIR LAS CLASES    
        <>
            <h3 className="animate__animated animate__fadeIn">{category}</h3>
            {loading && <p className="animate__animated animate__flash">Loading</p>}
            <div className='card-grid'>
                {
                    images.map(img => (
                        <GifGridItem 
                            key={img.id}
                            {...img}  //esto hace que las propiedades sean independientes
                        />
                    ))
                }
            </div>
        </>
    )
}

GifGrid.propType = {
    category: PropTypes.string.isRequired
}