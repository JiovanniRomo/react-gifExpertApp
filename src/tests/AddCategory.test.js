import { shallow } from 'enzyme';
import '@testing-library/jest-dom';
import React from 'react';
import { AddCategory } from '../components/AddCategory';


describe('Pruebas en <AddCategory />', () => {

    const setCategories = jest.fn();

    let wrapper = shallow(<AddCategory setCategory={setCategories}/>); 

    beforeEach(() => {
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategory={setCategories}/>); 
    });


    test('should correctly display the component', () => {

        expect(wrapper).toMatchSnapshot();
    });

    test('should change the text box ', () => {
        
        const input = wrapper.find('input');

        const value = 'Hola mundo'; //Esto es somo para simular el e.target.VALUE del handleInputChange, si no lo hicieramos, el value sería undefined

        input.simulate('change', { target: { value } }); //Aquí le asignamos el value al target

        expect(wrapper.find('p').text().trim()).toBe( value ); //Esto es solo para probar que funciona

    });

    test('shouldnt post the submit info', () => {
        
        wrapper.find('form').simulate('submit', {preventDefault(){}});
        
        expect(setCategories).not.toHaveBeenCalled();
    });

    test('should call setCategories and clean the text box', () => {
        
        //1.Simular el inputChange
        const value = 'Hola mundo';
        wrapper.find('input').simulate('change', {target: { value }});

        //2.Simular el submit
        wrapper.find('form').simulate('submit', {preventDefault(){}});

        //3 setCategories se debe haber llamado
        expect(setCategories).toHaveBeenCalled();
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function)); //esto evalua que se reciba cualquier tipo de FUNCIÓN

        //4 EL valor del input debe de estar ''
        expect(wrapper.find('input').prop('value')).toBe('');//Que el valor del input (obtenido desde los props) sea ''

    });

})