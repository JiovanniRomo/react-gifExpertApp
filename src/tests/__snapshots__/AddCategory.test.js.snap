// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> should correctly display the component 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
